File: python/easy/0819.mostCommonWord.py
import re
from typing import List


class Solution:
    def mostCommonWord(self, paragraph: str, banned: List[str]) -> str:
        word_count = {}
        split_word = re.findall(r"\w+", paragraph.lower())

        for word in split_word:
            if word not in banned:
                count = word_count.get(word, 0)
                word_count[word] = count + 1

        most_common = ""
        max_count = 0

        for word, count in word_count.items():
            if count > max_count:
                most_common = word
                max_count = count

        return most_common


if __name__ == "__main__":
    solution = Solution()
    print(
        solution.mostCommonWord(
            "Bob hit a ball, the hit BALL flew far after it was hit.", ["hit"]
        )
    )  # ball

File: typescript/easy/0819.mostCommonWord.ts
function mostCommonWord(paragraph: string, banned: string[]): string {
  const wordCount = new Map();
  const splitWord = paragraph.toLocaleLowerCase().match(/\w+/g) || [];

  for (let i = 0; i < splitWord.length; i++) {
    if (!banned.includes(splitWord[i])) {
      const count = wordCount.get(splitWord[i]) || 0;
      wordCount.set(splitWord[i], count + 1);
    }
  }

  let mostCommon = "";
  let maxCount = 0;

  for (const [word, count] of wordCount.entries()) {
    if (count > maxCount) {
      maxCount = count;
      mostCommon = word;
    }
  }

  return mostCommon;
}

console.log(mostCommonWord("Bob hit a ball, the hit BALL flew far after it was hit.", ["hit"])); // ball
